Core/MIPS/ARM64/Arm64IRCompVec.cpp

```diff
@@ -84,19 +84,322 @@ void Arm64JitBackend::CompIR_VecArith(IRInst inst) {
+
+	case Arm64Shuffle::EXT12_ZIP1_ADBA:
+		fp.EXT(EncodeRegToQuad(SCRATCHF1), vs, vs, 12);
+		fp.ZIP1(32, vd, vs, EncodeRegToQuad(SCRATCHF1));
+		return;
+
+	case Arm64Shuffle::DUP3_UZP1_DDAC:
+		fp.DUP(32, EncodeRegToQuad(SCRATCHF1), vs, 3);
+		fp.UZP1(32, vd, EncodeRegToQuad(SCRATCHF1), vs);
+		return;
+
+	default:
+		_assert_(false);
+		return;
+	}
+}
+
+uint8_t Arm64ShuffleResult(uint8_t mask, uint8_t prev) {
+	if (prev == 0xE4)
+		return mask;
+
+	uint8_t result = 0;
+	for (int i = 0; i < 4; ++i) {
+		int takeLane = (mask >> (i * 2)) & 3;
+		int lane = (prev >> (takeLane * 2)) & 3;
+		result |= lane << (i * 2);
+	}
+	return result;
+}
+
+int Arm64ShuffleScore(uint8_t shuf, uint8_t goal, int steps = 1) {
+	if (shuf == goal)
+		return 100;
+
+	int score = 0;
+	bool needs[4]{};
+	bool gets[4]{};
+	for (int i = 0; i < 4; ++i) {
+		uint8_t mask = 3 << (i * 2);
+		needs[(goal & mask) >> (i * 2)] = true;
+		gets[(shuf & mask) >> (i * 2)] = true;
+		if ((shuf & mask) == (goal & mask))
+			score += 4;
+	}
+
+	for (int i = 0; i < 4; ++i) {
+		if (needs[i] && !gets[i])
+			return 0;
+	}
+
+	// We need to look one level deeper to solve some, such as 1B (common) well.
+	if (steps > 0) {
+		int bestNextScore = 0;
+		for (int m = 0; m < (int)Arm64Shuffle::COUNT_NORMAL; ++m) {
+			uint8_t next = Arm64ShuffleResult(Arm64ShuffleMask((Arm64Shuffle)m), shuf);
+			int nextScore = Arm64ShuffleScore(next, goal, steps - 1);
+			if (nextScore > score) {
+				bestNextScore = nextScore;
+				if (bestNextScore == 100) {
+					// Take the earliest that gives us two steps, it's cheaper (not 2 instructions.)
+					score = 0;
+					break;
+				}
+			}
+		}
+
+		score += bestNextScore / 2;
+	}
+
+	return score;
+}
+
+Arm64Shuffle Arm64BestShuffle(uint8_t goal, uint8_t prev, bool needsCopy) {
+	// A couple special cases for optimal shuffles.
+	if (goal == 0x7C && prev == 0xE4)
+		return Arm64Shuffle::REV64_BADC;
+	if (goal == 0x2B && prev == 0xE4)
+		return Arm64Shuffle::EXT8_CDAB;
+	if ((goal == 0x07 || goal == 0x1C) && prev == 0xE4)
+		return Arm64Shuffle::EXT12_ZIP1_ADBA;
+	if ((goal == 0x8F || goal == 0x2F) && prev == 0xE4)
+		return Arm64Shuffle::DUP3_UZP1_DDAC;
+
+	// needsCopy true means insert isn't possible.
+	int attempts = needsCopy ? (int)Arm64Shuffle::COUNT_NOPREV : (int)Arm64Shuffle::COUNT_NORMAL;
+
+	Arm64Shuffle best = Arm64Shuffle::MOV_ABCD;
+	int bestScore = 0;
+	for (int m = 0; m < attempts; ++m) {
+		uint8_t result = Arm64ShuffleResult(Arm64ShuffleMask((Arm64Shuffle)m), prev);
+		int score = Arm64ShuffleScore(result, goal);
+		// Slightly discount options that involve an extra instruction.
+		if (m >= (int)Arm64Shuffle::COUNT_SIMPLE && m < (int)Arm64Shuffle::COUNT_NOPREV)
+			score--;
+		if (score > bestScore) {
+			best = (Arm64Shuffle)m;
+			bestScore = score;
+		}
+	}
+
+	_assert_(bestScore > 0);
+	return best;
+}
+
+
+static void Arm64ShufflePerform(ARM64FloatEmitter &fp, ARM64Reg vd, ARM64Reg vs, u8 shuf) {
```

Comment:
If for some reason this would start taking significant CPU time, it could build up a table of (shuf, vd != vs) -> sequence of (method, needsCopy), and then just look up in that table the next time. But yeah let's not bother now.